module/validation/receipt_validator.go

```diff
@@ -0,0 +1,183 @@
+	if err != nil {
+		return fmt.Errorf("failed to verify signature: %w", err)
+	}
+
+	if !valid {
+		return engine.NewInvalidInputErrorf("Invalid signature for (%x)", nodeIdentity.NodeID)
+	}
+
+	return nil
+}
+
+func (v *receiptValidator) verifyChunksFormat(result *flow.ExecutionResult) error {
+	for index, chunk := range result.Chunks.Items() {
+		if uint(index) != chunk.CollectionIndex {
+			return engine.NewInvalidInputErrorf("invalid CollectionIndex, expected %d got %d", index, chunk.CollectionIndex)
+		}
+
+		if chunk.BlockID != result.BlockID {
+			return engine.NewInvalidInputErrorf("invalid blockID, expected %v got %v", result.BlockID, chunk.BlockID)
+		}
+	}
+
+	// we create one chunk per collection, plus the
+	// system chunk. so we can check if the chunk number matches with the
+	// number of guarantees plus one; this will ensure the execution receipt
+	// cannot lie about having less chunks and having the remaining ones
+	// approved
+	requiredChunks := 1 // system chunk: must exist for block's ExecutionResult, even if block payload itself is empty
+
+	index, err := v.index.ByBlockID(result.BlockID)
+	if err != nil {
+		if !errors.Is(err, storage.ErrNotFound) {
+			return err
+		}
+		// reaching this line means the block is empty, i.e. it has no payload => we expect only the system chunk
+	} else {
+		requiredChunks += len(index.CollectionIDs)
+	}
+
+	if result.Chunks.Len() != requiredChunks {
+		return engine.NewInvalidInputErrorf("invalid number of chunks, expected %d got %d",
+			requiredChunks, result.Chunks.Len())
+	}
+
+	return nil
+}
+
+func (v *receiptValidator) verifyExecutionResult(result *flow.ExecutionResult) error {
+	prevResult, err := v.results.ByID(result.PreviousResultID)
+	if err != nil {
+		if errors.Is(err, storage.ErrNotFound) {
+			return engine.NewInvalidInputErrorf("receipt's previous result (%x) is unknown", result.PreviousResultID)
+		} else {
+			return err
+		}
+	}
+
+	block, err := v.state.AtBlockID(result.BlockID).Head()
+	if err != nil {
+		if errors.Is(err, storage.ErrNotFound) {
+			return engine.NewInvalidInputErrorf("no block found %v %w", result.BlockID, err)
+		} else {
+			return err
+		}
+	}
+
+	if prevResult.BlockID != block.ParentID {
+		return engine.NewInvalidInputErrorf("invalid block for previous result %v", prevResult.BlockID)
+	}
+
+	return nil
+}
+
+// Validate performs checks for ExecutionReceipt being valid or no.
+// Checks performed:
+// 	* can find stake and stake is positive
+//	* signature is correct
+//	* chunks are in correct format
+// 	* execution result has a valid parent
+// Returns nil if all checks passed successfully
+func (v *receiptValidator) Validate(receipt *flow.ExecutionReceipt) error {
+	identity, err := v.identityForNode(receipt.ExecutionResult.BlockID, receipt.ExecutorID)
+	if err != nil {
+		return err
+	}
+
+	err = v.ensureStakedNodeWithRole(identity, flow.RoleExecution)
+	if err != nil {
+		return err
```

Comment:
A good practice [is to wrap the error with more context](https://github.com/onflow/flow-go/blob/1efef2e885095890bcc9682b9638591fedcd30ba/state/protocol/badger/mutator.go#L93-L110), so that the error message could include the stack trace.